tests/test_model.py

```diff
@@ -103,10 +219,16 @@ def test_deserialize_error(self):
     def test_create(self):
```

Comment:
Is there a reason why we are not following the sample test code here from the tdd lab? Here's the sample test method given: 
```python
    def test_add_a_pet(self):
        """Create a pet and add it to the database"""
        pets = Pet.all()
        self.assertEqual(pets, [])
        pet = Pet(name="fido", category="dog", available=True, gender=Gender.Male)
        self.assertTrue(pet != None)
        self.assertEqual(pet.id, None)
        pet.create()
        # Asert that it was assigned an id and shows up in the database
        self.assertEqual(pet.id, 1)
        pets = Pet.all()
        self.assertEqual(len(pets), 1)
```
Also, doesn't the `MagicMock` overwrite the `db.session.add` and `db.session.commit` methods and therefore not actually testing if we can add an order to the database? It would be helpful if you could explain what we are testing here. Thank you.

Reply:
Is there a reason why we are not following the sample test code here from the tdd lab? 

The orders requirement is not same as pets. Order need to be tested on two different tables and different set of attributes which the sample code does not suffice. 
I want to practice writing unit tests with mocks. It seems to be appropriate in this case based on the lecture. The tests won't depend on database implementation so it will be deterministic.
Feel free to add any additional test case that you think need to be included in your patch.

Also, doesn't the MagicMock overwrite the db.session.add and db.session.commit methods and therefore not actually testing if we can add an order to the database? 
Yes however that is expected in this style. Database Session is external code as many external libraries are included. The unit tests should not be dependent on executing it. 
I prefer using mocks however I think we can have both types of test.

Reply:
- `order need to be tested on two different tables and different set of attributes which the sample code does`: Could you explain what you mean here? Not quite sure if I get it. Thank you : )
- `tests won't depend on database implementation so it will be deterministic`: My understanding was that even without the Mocks, the tests won't depend on database implementation that much, because SQLAlchemy abstracts that away. 
- `Yes however that is expected in this style. Database Session is external code as many external libraries are included.` But by doing this, if our code does not interact with SQLAlchemy properly, the tests won't be able to catch it. And I thought that is probably not what we want

Reply:
No, sqlachemy don't abstract it away. We are actually creating testdb and other stuff which is not exactly unit tests style.

I see sqlachemy has a mock library to abstract it "alchemy-mock · PyPI" https://pypi.org/project/alchemy-mock/" that abstract some of it. So decided to do it this way.
 Database Session is external code as many external libraries are included. But by doing this, if our code does not interact with SQLAlchemy properly?The tests won't be able to catch it. And I thought that is probably not what we want

I follow the style were we test that sqlachemy part is getting "called" correctly thus interaction with sqlachemy is tested. As long as the sql alchemy functions are called , we don't care if the sql alchemy implementation is correct  as that is external code.

Reply:
> No, sqlachemy don't abstract it away. We are actually creating testdb and other stuff which is not exactly unit tests style.
> 
> I see sqlachemy has a mock library to abstract it "alchemy-mock · PyPI" https://pypi.org/project/alchemy-mock/" that abstract some of it. So decided to do it this way.
> Database Session is external code as many external libraries are included. But by doing this, if our code does not interact with SQLAlchemy properly?The tests won't be able to catch it. And I thought that is probably not what we want
> 
> I follow the style were we test that sqlachemy part is getting "called" correctly thus interaction with sqlachemy is tested. As long as the sql alchemy functions are called , we don't care if the sql alchemy implementation is correct as that is external code.

Mocking something and abstracting away something are two different things. SQLAlchemy abstracts away the database functionalities because we're writing Python and not SQL to interact with the database. 

SQLAlchemy functions being called doesn't mean they were called correctly or necessarily did what we wanted them to do. Plus, the professor's sample TTD code indeed created a test database and used it for testing.

Reply:
Sqlachemy also have logic to convert python part to sql. I don't think good unit test should run that otherwise it will become integration test.
For example this post
"The Difference Between Unit, Integration and Functional Testing" https://www.softwaretestinghelp.com/the-difference-between-unit-integration-and-functional-testing/amp/

Mentions, unit test should not test any dependencies. (Sqlachemy in this case). Imagine if there is a bug in sqlachemy in the next version still unit test shouldn't fail. 

Using testdb is actually integration testing based on my exploration on the topic. I think both are useful.  As mentioned earlier you can  add additional tests in your patch.

Reply:
> Sqlachemy also have logic to convert python part to sql. I don't think good unit test should run that otherwise it will become integration test.
> For example this post
> "The Difference Between Unit, Integration and Functional Testing" https://www.softwaretestinghelp.com/the-difference-between-unit-integration-and-functional-testing/amp/
> 
> Mentions, unit test should not test any dependencies. (Sqlachemy in this case). Imagine if there is a bug in sqlachemy in the next version still unit test shouldn't fail.
> 
> Using testdb is actually integration testing based on my exploration on the topic. I think both are useful. As mentioned earlier you can add additional tests in your patch.

Yeah I think your point is totally valid. And thank you for doing the research. However, since the sample code we were told to follow (the TDD lab) contradicts it, I think it might be to everyone's benefit to follow the template code.

Reply:
Mock doesn't contradicts and aligned with professor lectures. Let us do that from next time after discussing this with Prof. I am okay with other tests being added that follow sample code so in the interest of time let us keep them in this pull request.